src/librustc_mir/borrow_check/error_reporting.rs

```diff
@@ -1060,16 +1066,118 @@ enum StorageDeadOrDrop<'tcx> {
+             target={:?}",
+             location, place, target,
+        );
+        for stmt in &self.mir[location.block].statements[location.statement_index..] {
+            debug!(
+                "add_closure_invoked_twice_with_moved_variable_suggestion: stmt={:?} \
+                 target={:?}",
+                 stmt, target,
+            );
+            if let StatementKind::Assign(into, box Rvalue::Use(from)) = &stmt.kind {
+                debug!(
+                    "add_closure_invoked_twice_with_moved_variable_suggestion: into={:?} \
+                     from={:?}",
+                     into, from,
+                );
+                match from {
+                    Operand::Copy(ref place) |
+                    Operand::Move(ref place) if target == place.local() =>
+                        target = into.local(),
+                    _ => {},
+                }
+            }
+        }
+
+
+        let terminator = self.mir[location.block].terminator();
+        debug!(
+            "add_closure_invoked_twice_with_moved_variable_suggestion: terminator={:?}",
+            terminator,
+        );
+        if let TerminatorKind::Call {
+            func: Operand::Constant(box Constant {
+                literal: ty::Const { ty: &ty::TyS { sty: ty::TyKind::FnDef(id, _), ..  }, ..  },
+                ..
+            }),
+            args,
+            ..
+        } = &terminator.kind {
+            debug!("add_closure_invoked_twice_with_moved_variable_suggestion: id={:?}", id);
+            if self.infcx.tcx.parent(id) == self.infcx.tcx.lang_items().fn_once_trait() {
+                let closure = match args.first() {
+                    Some(Operand::Copy(ref place)) |
+                    Some(Operand::Move(ref place)) if target == place.local() =>
+                        place.local().unwrap(),
+                    _ => return,
+                };
+                debug!(
+                    "add_closure_invoked_twice_with_moved_variable_suggestion: closure={:?}",
+                     closure,
+                );
+
+                if let ty::TyKind::Closure(did, substs) = self.mir.local_decls[closure].ty.sty {
+                    let upvar_tys = substs.upvar_tys(did, self.infcx.tcx);
+                    let node_id = match self.infcx.tcx.hir.as_local_node_id(did) {
+                        Some(node_id) => node_id,
+                        _ => return,
+                    };
+
+                    self.infcx.tcx.with_freevars(node_id, |freevars| {
+                        for (freevar, upvar_ty) in freevars.iter().zip(upvar_tys) {
+                            debug!(
+                                "add_closure_invoked_twice_with_moved_variable_suggestion: \
+                                 freevar={:?} upvar_ty={:?}",
+                                freevar, upvar_ty,
+                            );
+                            if !upvar_ty.is_region_ptr() {
```

Comment:
This is too broad -- the code is only looking at the type of the value that was captured (and crudely at that), but it needs to take into consideration *how* the value is used (e.g., the `&*x` in the `issue-12127.rs` does not require moving `x`).

There are a few ways to handle this.

One way would be to go through the point where the closure is **created** and look at whether it "moves" its arguments. That seems a bit complex though.

Another way would be to consult the [`closure_kind_origins` map in the typeck-tables](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TypeckTables.html#structfield.closure_kind_origins). This map indicates *why* a given closure has the kind it does (and in particular it will have an entry if this is due to a user of some freevar).

See the code in `librustc_borrowck` for an example:

https://github.com/rust-lang/rust/blob/5ea8eb55cd9f4547b332f43c9f723de30187c223/src/librustc_borrowck/borrowck/mod.rs#L704-L713

(I think that in the case of issue-12127, there will be no entry in the map at all.)

Reply:
Pushed a fix for this.